packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -73,44 +64,227 @@ class Intent with Diagnosticable {
+              'The $runtimeType sending notification was',
+              this,
+              style: DiagnosticsTreeStyle.errorProperty,
+            );
+          },
+        ));
+      }
+    }
+  }
+}
+
+/// A helper widget for making sure that listeners on an action are removed properly.
+///
+/// Listeners on the [Action] class must have their listener callbacks removed
+/// with [Action.removeActionListener] when the listener is disposed of. This widget
+/// helps with that, by providing a lifetime for the connection between the
+/// [listener] and the [Action], and by handling the adding and removing of
+/// the [listener] at the right points in the widget lifecycle.
+///
+/// If you listen to an [Action] widget in a widget hierarchy, you should use
+/// this widget. If you are using an [Action] outside of a widget context, then
+/// you must call removeListener yourself.
+@immutable
+class ActionListener extends StatefulWidget {
+  /// Create a const [ActionListener].
+  ///
+  /// The [listener], [action], and [child] arguments must not be null.
+  const ActionListener({
+    Key key,
+    @required this.listener,
+    @required this.action,
+    @required this.child,
+  })  : assert(listener != null),
+        assert(action != null),
+        assert(child != null),
+        super(key: key);
+
+  /// The [ActionListenerCallback] callback to register with the [action].
+  ///
+  /// Must not be null.
+  final ActionListenerCallback listener;
+
+  /// The [Action] that the callback will be registered with.
+  ///
+  /// Must not be null.
+  final Action<Intent> action;
+
+  /// {@macro flutter.widgets.child}
+  final Widget child;
 
   @override
-  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
-    super.debugFillProperties(properties);
-    properties.add(DiagnosticsProperty<LocalKey>('intentKey', intentKey));
+  _ActionListenerState createState() => _ActionListenerState();
+}
+
+class _ActionListenerState extends State<ActionListener> {
+  @override
+  void initState() {
+    super.initState();
+    widget.action.addActionListener(widget.listener);
   }
+
+  @override
+  void didUpdateWidget(ActionListener oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.action == widget.action && oldWidget.listener == widget.listener) {
+      return;
+    }
+    oldWidget.action.removeActionListener(oldWidget.listener);
+    widget.action.addActionListener(widget.listener);
+  }
+
+  @override
+  void dispose() {
+    widget.action.removeActionListener(widget.listener);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) => widget.child;
+}
+
+/// An abstract [Action] subclass that adds an optional [BuildContext] to the [invoke]
+/// method to be able to provide context to actions.
+///
+/// [ActionDispatcher.invokeAction] checks to see if the action it is invoking
+/// is a [ContextAction], and if it is, tries to supply it with a context.
```

Comment:
"tries to" makes me wonder what happens if it can't and what the situations are where it can't. Maybe document that?

Reply:
Or just take out the "tries to supply" and replace with "supplies".  It always does if it's a ContextAction. Fixed.